k2/csrc/fsa_utils.cu

```diff
@@ -0,0 +1,327 @@
+    auto sub = in.substr(start);
+    TrimString(&sub);
+    if (!sub.empty()) out->emplace_back(std::move(sub));
+  }
+}
+
+/* Check that arcs are valid.
+
+   It will abort the program if there is an invalid arc.
+   An arc is invalid if one of the following happens:
+
+    - its symbol is -1, but its dest state is not final state.
+    - its dest state is larger than final state
+
+  @param [in] state_arcs   Indexed by state numbers.
+  @param [in] final_state  The final state.
+ */
+static void CheckStateArcs(const std::vector<std::vector<Arc>> &state_arcs,
+                           int32_t final_state) {
+  int32_t num_states = static_cast<int32_t>(state_arcs.size());
+  K2_CHECK_EQ(num_states, final_state + 1);
+  K2_CHECK(!state_arcs[final_state].empty());
+
+  for (int32_t i = 0; i != num_states; ++i) {
+    const auto &arcs = state_arcs[i];
+    for (const auto &arc : arcs) {
+      K2_CHECK_EQ(arc.src_state, i);
+      K2_CHECK_LE(arc.dest_state, final_state);
+      if (arc.symbol == -1) K2_CHECK_EQ(arc.dest_state, final_state);
+    }
+  }
+}
+
+/* Build an Fsa from a list of lists of arcs.
+
+  @param [in] state_arcs Indexed by states.
+  @param [in] num arcs   It is the number of arcs in total.
+                         Provided for optimization.
+ */
+static Fsa AcceptorFromStateArcs(
+    const std::vector<std::vector<Arc>> &state_arcs, int32_t num_arcs) {
+  int32_t num_states = static_cast<int32_t>(state_arcs.size());
+  std::vector<int32_t> row_splits(num_states + 1);
+  std::vector<Arc> arcs;
+  arcs.reserve(num_arcs);
+
+  row_splits[0] = 0;
+  for (int32_t i = 1; i != num_states; ++i) {
+    row_splits[i] =
+        row_splits[i - 1] + static_cast<int32_t>(state_arcs[i - 1].size());
+    arcs.insert(arcs.end(), state_arcs[i - 1].begin(), state_arcs[i - 1].end());
+  }
+  // the final_state has no leaving arcs, so there is nothing
+  // to add to `arcs` and row_splits[num_states]
+  row_splits[num_states] = row_splits[num_states - 1];
+
+  auto cpu_context = GetCpuContext();
+  Array1<int32_t> _row_splits(cpu_context, row_splits);
+  std::vector<RaggedShapeDim> axes(1);
+  axes[0].row_splits = _row_splits;
+  axes[0].cached_tot_size = -1;
+
+  RaggedShape shape(axes);
+  Array1<Arc> values(cpu_context, arcs);
+
+  return Fsa(shape, values);
+}
+
+// Create an acceptor from a stream.
+static Fsa AcceptorFromStream(std::string first_line, std::istringstream &is,
+                              bool negate_scores) {
+  std::vector<std::vector<Arc>> state_arcs;  // indexed by states
+  std::string line = std::move(first_line);
+  std::vector<std::string> splits;
+
+  float scale = 1;
+  if (negate_scores) scale = -1;
+
+  int32_t final_state = -1;  // invalid state
+  int32_t num_arcs = 0;
+  do {
+    SplitStringToVector(line, kDelim,
+                        &splits);  // splits is cleared in the function
+    if (splits.empty()) continue;  // this is an empty line
+
+    auto num_fields = splits.size();
```

Comment:
Actually it should be possible to just read the arcs as a vector and construct the FSA from that, based on existing functions in fsa.h.  (My other PR will correct some issues in their implementations).  But we can leave this for now, and involve that other code later after this is tested.

Reply:
Fixed. Replaced with API `FsaFromArray1`.